File: AutoCompleteEngine/Python/ranking.py
from typing import List
from models import WordEntry
import abc

class RankingStrategy(abc.ABC):
    """
    Abstract Base Class for Ranking Strategies.
    """
    @abc.abstractmethod
    def rank_entries(self, entries: List[WordEntry], prefix: str) -> List[WordEntry]:
        pass

class FrequencyStrategy(RankingStrategy):
    """
    Sorts based on Usage Count (Descending).
    """
    def rank_entries(self, entries: List[WordEntry], prefix: str) -> List[WordEntry]:
        # Sort by usage_count desc
        return sorted(entries, key=lambda x: x.usage_count, reverse=True)

class RelevanceStrategy(RankingStrategy):
    """
    Sorts based on EntityType Score (Descending), then length of word (shorter is better).
    """
    def rank_entries(self, entries: List[WordEntry], prefix: str) -> List[WordEntry]:
        # Sort by Entity Score desc, then by Word Length asc
        return sorted(entries, key=lambda x: (-x.entity_type.score, len(x.word)))

File: AutoCompleteEngine/Java/ranking.java
// Strategy interface for ranking algorithms

import java.util.Comparator;

interface RankingStrategy {
    List<String> rank(List<String> suggestions, String prefix);
}

// Concrete strategy: Rank by relevance (simple substring match)
class RelevanceStrategy implements RankingStrategy {
    @Override
    public List<String> rank(List<String> suggestions, String prefix) {
        suggestions.sort((a, b) -> {
            int compareScore = Double.compare(b.EntityType.getScore(), a.EntityType.getScore());
            if (compareScore != 0) {
                return compareScore;
            }
            return Integer.compare(a.word.length(), b.word.length());
        });
        return suggestions;
    }
}

class FrequencyStrategy implements RankingStrategy {
    @Override
    public List<String> rank(List<String> suggestions, String prefix) {
        suggestions.sort((a, b) -> {
            return Integer.compare(b.usageCount(), a.usageCount());
        });
        return suggestions;
    }
}